File: appleApp/Sources/RallyCore/StreamRequestHeaders.swift
import Foundation

/// 1:1 port of `sanitizedStreamHeaders` + `normalizedBearerToken` in
/// `presentation/player/StreamRequestHeaders.kt`.
public enum StreamRequestHeaders {
    private static let allowed: Set<String> = [
        "accept", "accept-language", "authorization", "cookie", "origin", "referer", "user-agent",
    ]

    public static func sanitized(_ headers: [String: String]?) -> [String: String] {
        var out: [String: String] = [:]
        for (name, value) in headers ?? [:] {
            guard allowed.contains(name.lowercased()),
                  name.count <= 64, value.count <= 4096,
                  !containsNewline(name), !containsNewline(value) else { continue }
            out[name] = value
        }
        return out
    }

    /// Scalar-level check: Swift grapheme `contains("\n")` misses a CR+LF
    /// pair (one cluster), so inspect scalars like the Kotlin Char check.
    private static func containsNewline(_ value: String) -> Bool {
        value.unicodeScalars.contains { $0.value == 10 || $0.value == 13 }
    }

    public static func normalizedBearerToken(_ token: String) -> String {
        token.lowercased().hasPrefix("bearer ") ? token : "Bearer \(token)"
    }
}

File: windowsApp/src/Rally.Core/StreamRequestHeaders.cs
// 1:1 port of sanitizedStreamHeaders + normalizedBearerToken in
// presentation/player/StreamRequestHeaders.kt.
namespace Rally.Core;

public static class StreamRequestHeaders
{
    private static readonly HashSet<string> Allowed = new(StringComparer.OrdinalIgnoreCase)
        { "accept", "accept-language", "authorization", "cookie", "origin", "referer", "user-agent" };

    public static Dictionary<string, string> Sanitize(Dictionary<string, string>? headers)
    {
        var out_ = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (headers is null) return out_;
        foreach (var (name, value) in headers)
        {
            if (!Allowed.Contains(name)) continue;
            if (name.Length > 64 || value.Length > 4096) continue;
            // Ordinal (code-unit) check: catches lone CR, lone LF, and CR+LF pairs.
            // (Swift needs a scalar check here; C# ordinal Contains is sufficient.)
            if (name.Contains('\n') || name.Contains('\r')) continue;
            if (value.Contains('\n') || value.Contains('\r')) continue;
            out_[name] = value;
        }
        return out_;
    }

    public static string NormalizedBearerToken(string token) =>
        token.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase) ? token : "Bearer " + token;
}
